### PR TITLE
Migrate to swift 3

### PR DESCRIPTION
**Pull request by typfel**
## What's in this PR
- Migrate to swift 3
- Turn on extension API only flag

We should also investigate if we could improve the error type bridging.

### CODE REVIEW

File: Cryptobox.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -653,10 +656,12 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = 0928E3741BA17E300057232E /* Cryptobox-ios.xcconfig */;
 			buildSettings = {
+				ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES = YES;
```
Comment (mikeger) [inline Cryptobox.xcodeproj/project.pbxproj]: Please check that swift frameworks are NOT embedded into the final .framework.

Comment (marcoconti83) [inline Cryptobox.xcodeproj/project.pbxproj]: The build script takes care of that, it has a step called "... otherwise Apple will fail submission"

Comment (mikeger) [inline Cryptobox.xcodeproj/project.pbxproj]: Yep, but if you build with Carthage then the folder would be there. I think the project can be configured not to include extra nested frameworks (or may be apple fixed it?).

File: CryptoboxTests/EncryptionSessionsDirectoryTests.swift [hunk 2]
```diff
@@ -63,60 +63,60 @@ extension EncryptionSessionsDirectoryTests {
...
         /// This will force commit
         self.recreateStatuses()
     }
 
 
     enum Person {
-        case Alice
-        case Bob
+        case alice
+        case bob
```
Comment (mikeger) [inline CryptoboxTests/EncryptionSessionsDirectoryTests.swift]: Potentially it should be possible to have upper-case enum members? It's names, I think they should be upper-case.

**Comment by mikeger** [comment]
Looks good to me!

File: CryptoboxTests/EncryptionSessionsDirectoryTests.swift (hunk 2 above)
Comment (daehn) [inline CryptoboxTests/EncryptionSessionsDirectoryTests.swift]: It's still possible, just the Swift 3 naming guidelines recommend using lowercase values

File: Cryptobox/CryptoBoxError.swift [hunk 3]
```diff
@@ -19,79 +19,79 @@
...
 
-extension CBoxResult : ErrorType {
+extension CBoxResult : Error {
 }
 
-@objc public enum CryptoboxError : UInt32, ErrorType {
+@objc public enum CryptoboxError : UInt32, Error {
     
-    case CryptoboxSuccess                 = 0
+    case cryptoboxSuccess                 = 0
```
Comment (daehn) [inline Cryptobox/CryptoBoxError.swift]: We could remove the redundant `cryptobox` prefix on every enum case

File: Cryptobox/EncryptionContext.swift [hunk 4]
```diff
@@ -57,40 +57,40 @@ class _CBox : PointerWrapper {}
...
+        case save
+        case discard
     }
     
     /// Underlying C-style implementation
     let implementation = _CBox()
     
     /// File directory with the implementation files
-    private let path : NSURL
+    fileprivate let path : URL
```
Comment (daehn) [inline Cryptobox/EncryptionContext.swift]: Could these `fileprivate` methods also be `private`?

File: CryptoboxTests/EncryptionSessionsDirectoryTests.swift (hunk 2 above)
Comment (typfel) [inline CryptoboxTests/EncryptionSessionsDirectoryTests.swift]: OK I'll make the names uppercase

File: Cryptobox/EncryptionContext.swift (hunk 4 above)
Comment (typfel) [inline Cryptobox/EncryptionContext.swift]: I'll try

Comment (typfel) [inline Cryptobox/EncryptionContext.swift]: they need to be `fileprivate` since they are accessed in extensions.

**Comment by typfel** [comment]
@marcoconti83 @mikeger @daehn answered  & fixed PR comments.

**Comment by mikeger** [comment]
@typfel please create binary release

**Pull request merged**